xlators/mgmt/glusterd/src/glusterd-svc-helper.c

```diff
@@ -862,65 +861,59 @@ __glusterd_send_svc_configure_req(glusterd_svc_t *svc, int flags,
     req_size = xdr_sizeof((xdrproc_t)xdr_gd1_mgmt_brick_op_req, req);
     iobuf = iobuf_get2(rpc->ctx->iobuf_pool, req_size);
     if (!iobuf) {
-        goto *errlbl;
+        goto err;
     }
-    errlbl = &&maybe_free_iobuf;
 
     iov.iov_base = iobuf->ptr;
     iov.iov_len = iobuf_pagesize(iobuf);
 
     iobref = iobref_new();
     if (!iobref) {
-        goto *errlbl;
+        goto err;
     }
-    errlbl = &&free_iobref;
 
     iobref_add(iobref, iobuf);
-    /*
-     * Drop our reference to the iobuf.  The iobref should already have
-     * one after iobref_add, so when we unref that we'll free the iobuf as
-     * well.  This allows us to pass just the iobref as frame->local.
-     */
-    iobuf_unref(iobuf);
-    /* Set the pointer to null so we don't free it on a later error. */
-    iobuf = NULL;
 
     /* Create the xdr payload */
     ret = xdr_serialize_generic(iov, req, (xdrproc_t)xdr_gd1_mgmt_brick_op_req);
     if (ret == -1) {
-        goto *errlbl;
+        goto err;
     }
     iov.iov_len = ret;
 
     /* Send the msg */
     GF_ATOMIC_INC(conf->blockers);
     ret = rpc_clnt_submit(rpc, &gd_brick_prog, op, cbkfn, &iov, 1, NULL, 0,
                           iobref, frame, NULL, 0, NULL, 0, NULL);
-    if (dict)
-        dict_unref(dict);
-    GF_FREE(volfile_content);
-    if (spec_fd >= 0)
-        sys_close(spec_fd);
-    return ret;
 
-free_iobref:
-    iobref_unref(iobref);
-maybe_free_iobuf:
+    /* Checking if the rpc_clnt_submit() failed beacuse of memory allocation,
+     * which implies the frame is not destroyed there. So, it needs to be
+     * destroyed here */
+    if (ret != -2)
+        frame = NULL;
+
+    /* Changing ret back to -1 so as to avoid -2 being passed to caller,
+     * considering that its not the usual error number used in the code */
+    if (ret == -2)
+        ret = -1;
+err:
     if (iobuf) {
         iobuf_unref(iobuf);
     }
-err:
+    if (iobref) {
+        iobref_unref(iobref);
+    }
     if (dict)
         dict_unref(dict);
-    if (brick_req.dict.dict_val)
+    if (ret && brick_req.dict.dict_val)
         GF_FREE(brick_req.dict.dict_val);
 
     GF_FREE(volfile_content);
     if (spec_fd >= 0)
         sys_close(spec_fd);
-    if (frame)
+    if (ret && frame)
```

Comment:
Checking `err` here seems redundant. When `frame` needs to be preserved, it's already set to `NULL`.

Reply:
I didn't understand this part. You meant to say that since we are already assigning `frame` as NULL, we don't need to check this condition. Is it so?

Reply:
If that's so, then it's just one condition and I don't think it will do much harm. But it can be improved by changing it to `if(frame && ret)`. Otherwise, we would have to add another label and handle it, which I think would introduce unnecessary complexity.( Just a thought)